scripts/keyless/keylessDeploy.ts

```diff
@@ -0,0 +1,293 @@
+        "0xf8a58085174876e800830186a08080b853604580600e600039806000f350fe7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf31ba02222222222222222222222222222222222222222222222222222222222222222a02222222222222222222222222222222222222222222222222222222222222222",
+      );
+    } catch (err) {
+      console.log("Couldn't deploy factory");
+      console.log(err);
+      process.exit(1);
+    }
+
+    // check
+    factoryCode = await hardhatEthers.provider.getCode(commonFactory);
+    console.log("deployed code: ", factoryCode);
+  }
+}
+
+export async function constructKeylessTx(bytecode: any, args: any) {
+  const bytecodeHash = hardhatEthers.utils.id(bytecode);
+  const salt = `tw.${bytecodeHash}`;
+  const saltHash = hardhatEthers.utils.id(salt);
+
+  const data = hardhatEthers.utils.solidityPack(["bytes32", "bytes", "bytes"], [saltHash, bytecode, args]);
+
+  const testTx = {
+    gasPrice: 100 * 10 ** 9,
+    gasLimit: 5000000,
+    to: commonFactory,
+    value: 0,
+    nonce: 0,
+    data: data,
+  };
+  const customSignature = ethers.utils.joinSignature(customSigInfo);
+  const serializedTestTx = ethers.utils.serializeTransaction(testTx);
+
+  const addr = ethers.utils.recoverAddress(
+    ethers.utils.arrayify(ethers.utils.keccak256(serializedTestTx)),
+    customSignature,
+  );
+
+  const signedSerializedTestTx = ethers.utils.serializeTransaction(testTx, customSigInfo);
+
+  return {
+    addr,
+    signedSerializedTestTx,
+  };
+}
+
+export async function computeDeploymentAddress(bytecode: any, args: any) {
+  const bytecodeHash = hardhatEthers.utils.id(bytecode);
+  const salt = `tw.${bytecodeHash}`;
+  const saltHash = hardhatEthers.utils.id(salt);
+
+  const deployData = hardhatEthers.utils.solidityPack(["bytes32", "bytes", "bytes"], [saltHash, bytecode, args]);
+
+  const initBytecode = hardhatEthers.utils.solidityPack(["bytes", "bytes"], [bytecode, args]);
+  const deployInfoPacked = hardhatEthers.utils.solidityPack(
+    ["bytes1", "address", "bytes32", "bytes32"],
+    ["0xff", commonFactory, saltHash, hardhatEthers.utils.solidityKeccak256(["bytes"], [initBytecode])],
+  );
+  const addr = hardhatEthers.utils.solidityKeccak256(["bytes"], [deployInfoPacked]);
+
+  return { predictedAddress: `0x${addr.slice(26)}`, deployData };
+}
+
+async function deployInfraKeyless() {
+  const [signer] = await hardhatEthers.getSigners();
+
+  // create2 factory
+  await deployCommonFactory(signer);
+
+  for (let txInfo of Object.values(infraContracts) as InfraTxInfo[]) {
+    const code = await hardhatEthers.provider.getCode(txInfo.predictedAddress);
+    if (code === "0x") {
+      let fundAddr = {
+        to: txInfo.from,
+        value: hardhatEthers.utils.parseEther("1"),
+      };
+      await signer.sendTransaction(fundAddr);
+
+      await (await hardhatEthers.provider.sendTransaction(txInfo.tx)).wait();
```

Comment:
this is a beauty. btw i think we should be able to estimate gas simply by calling `provider.estimateGas(tx)` here to get an accurate amount to send